source/develop/release-management/features/infra/link-following.html.md

```diff
@@ -0,0 +1,114 @@
+---
+title: API Link Following
+category: feature
+authors: oliel
+wiki_category: Feature
+wiki_title: API Link Following
+wiki_revision_count: 0
+wiki_last_updated: 2017-7-7
+feature_name: Link Following
+feature_modules: api (engine)
+feature_status: Released
+---
+
+## Feature Name
+
+API Link Following
+
+## Summary
+
+This features enable oVirt API users to request that the contents of some of the entity's links be returned inline, inside the requested entity.
+
+## Owner
+
+*   Name: Ori Liel (oliel)
+
+*   Email: <oliel@redhat.com>
+
+## Detailed Description
+
+Motivation:
+
+Currently when there is the need to retrieve multiple related objects from the API the only alternative is to retrieve the first one, and then, send additional requests to retrieve the related objects. For example, if you need a virtual machine and also the disks and NICs you need first to send a request like this:
+
+  <code>GET /ovirt-engine/api/vms/123</code>
+
+And then additional requests to get the disk attachments, the disks, and the NICs:
+
+  <code>GET /ovirt-engine/api/vms/123/diskattachments</code>
+  <code>GET /ovirt-engine/api/disks/456</code>
+  <code>GET /ovirt-engine/api/disks/789</code>
+  <code>GET /ovirt-engine/api/vms/123/nics</code>
+
+In an environment with high latency this multiplies the time required to retrieve the data. In addition it also means that multiple queries have to be sent to the database to retrieve the data.
+
+In order to improve in these two areas the new <code>follow</code> parameter will be introduced. This parameter will be a list of links that the server should follow and populate. For example, the previous scenario will be solved sending this request:
+
+  <code>GET /ovirt-engine/api/vms/123?follow=diskattachments.disks,nics</code>
+
+That will return the virtual machine with the disks and the NICs embedded in the same response, thus avoiding the multiple network round-trips.
+
+The multiple database queries will be avoided only if the server is modified to retrieve that data with more efficient queries, otherwise the server will use the naive approach of calling itself to retrieve it, which won’t improve the number of queries.
+
+The first step is analyzing the value of follow and using it to fetch the desired hrefs - the actual links to be followed. Single entities contain their href directly inside them (e.g: <code>vm.getCluster().getHref())</code>, and hrefs of sub-collections (e.g: nics, disks) are stored in the links collection (1. vm.getLinks() 2. Find specific link. 3.link.getHref()). This division exists for historic reasons.
```

Comment:
can you check brackets in the above paragraph

Reply:
What do you mean check brackets?